File: szavazas-core/src/main/java/hu/kdea/szavazas/ballotprocessor/aruco/BoofCVArucoDetector.java
package hu.kdea.szavazas.ballotprocessor.aruco;

import boofcv.abst.fiducial.FiducialDetector;
import boofcv.alg.distort.RemovePerspectiveDistortion;
import boofcv.alg.geo.h.HomographyDirectLinearTransform;
import boofcv.factory.fiducial.ConfigHammingMarker;
import boofcv.factory.fiducial.FactoryFiducial;
import boofcv.factory.fiducial.HammingDictionary;
import boofcv.struct.geo.AssociatedPair;
import boofcv.struct.image.GrayF32;
import boofcv.struct.image.GrayU8;
import boofcv.struct.image.ImageType;
import boofcv.struct.image.Planar;
import georegression.struct.homography.Homography2D_F64;
import georegression.struct.homography.UtilHomography_F64;
import georegression.struct.point.Point2D_F64;
import georegression.struct.shapes.Polygon2D_F64;
import georegression.transform.homography.HomographyPointOps_F64;
import hu.kdea.szavazas.ballotprocessor.Logger;
import org.ejml.data.DMatrixRMaj;
import kotlin.Pair;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class BoofCVArucoDetector implements IArucoDetector {
    private static final int TOP_LEFT_ID = 37;
    private static final int TOP_RIGHT_ID = 50;
    private static final int BOTTOM_LEFT_ID = 44;
    private static final int BOTTOM_RIGHT_ID = 219;
    private static final Set<Integer> REQUIRED_IDS = new HashSet<>(Set.of(TOP_LEFT_ID, TOP_RIGHT_ID, BOTTOM_LEFT_ID, BOTTOM_RIGHT_ID));

    private Point2D_F64 bottomLeftTopOriginal;
    private Point2D_F64 bottomRightTopOriginal;
    private Homography2D_F64 homography;

    @Override
    public List<Point2D_F64> findBallotCorners(GrayU8 gray) {
        Map<Integer, List<Point2D_F64>> markers = scanMarkers(gray);
        if (markers == null) {
            return null;
        }
        captureBottomMidpoints(markers);
        Logger.INSTANCE.d("ArUco", "Ballot corners detected successfully (IDs: " + markers.keySet() + ")");
        return extractOuterCorners(markers);
    }

    @Override
    public Planar<GrayU8> warpBallot(Planar<GrayU8> src, List<Point2D_F64> srcPoints) {
        Pair<Integer, Integer> outputSize = computeOutputSize(srcPoints);
        int width = outputSize.getFirst();
        int height = outputSize.getSecond();
        homography = computeHomography(srcPoints, width, height);
        Planar<GrayF32> srcF32 = convertU8ToF32(src);
        Planar<GrayF32> outF32 = applyPerspective(srcF32, srcPoints, width, height);
        return convertF32ToU8(outF32, width, height);
    }

    @Override
    public Double bottomMarkerTopInScaled(double scaleFactor) {
        if (homography == null || bottomLeftTopOriginal == null || bottomRightTopOriginal == null) {
            return null;
        }
        double ty1 = HomographyPointOps_F64.transform(homography, bottomLeftTopOriginal, null).y;
        double ty2 = HomographyPointOps_F64.transform(homography, bottomRightTopOriginal, null).y;
        return Math.min(ty1, ty2) * scaleFactor;
    }

    private Map<Integer, List<Point2D_F64>> scanMarkers(GrayU8 gray) {
        FiducialDetector<GrayU8> detector = createDetector();
        detector.detect(gray);
        Logger.INSTANCE.d("ArUco", "Total markers found: " + detector.totalFound());
        Map<Integer, List<Point2D_F64>> markers = new HashMap<>();
        for (int i = 0; i < detector.totalFound(); i++) {
            Pair<Integer, List<Point2D_F64>> marker = readMarker(detector, i);
            int id = marker.getFirst();
            List<Point2D_F64> corners = marker.getSecond();
            logMarker(id, corners);
            if (REQUIRED_IDS.contains(id) && markers.containsKey(id)) {
                Logger.INSTANCE.d("ArUco", "Duplicate required marker id " + id + " – failing");
                return null;
            }
            markers.put(id, corners);
        }
        if (!markers.keySet().containsAll(REQUIRED_IDS)) {
            Logger.INSTANCE.d("ArUco", "Missing required markers. Found IDs: " + markers.keySet());
            return null;
        }
        return markers;
    }

    private FiducialDetector<GrayU8> createDetector() {
        ConfigHammingMarker config = ConfigHammingMarker.loadDictionary(HammingDictionary.ARUCO_MIP_16h3);
        return FactoryFiducial.squareHamming(config, null, GrayU8.class);
    }

    private Pair<Integer, List<Point2D_F64>> readMarker(FiducialDetector<GrayU8> detector, int index) {
        int id = (int) detector.getId(index);
        Polygon2D_F64 bounds = new Polygon2D_F64();
        detector.getBounds(index, bounds);
        List<Point2D_F64> corners = new ArrayList<>();
        for (int i = 0; i < bounds.size(); i++) {
            corners.add(bounds.get(i));
        }
        return new Pair<>(id, corners);
    }

    private void logMarker(int id, List<Point2D_F64> corners) {
        double sumX = 0.0;
        double sumY = 0.0;
        for (Point2D_F64 corner : corners) {
            sumX += corner.x;
            sumY += corner.y;
        }
        int cx = (int) (sumX / corners.size());
        int cy = (int) (sumY / corners.size());
        Logger.INSTANCE.d("ArUco", "  id=" + id + "  center=(" + cx + "," + cy + ")");
    }

    private List<Point2D_F64> extractOuterCorners(Map<Integer, List<Point2D_F64>> markers) {
        return List.of(
            markers.get(TOP_LEFT_ID).get(1),
            markers.get(TOP_RIGHT_ID).get(2),
            markers.get(BOTTOM_RIGHT_ID).get(3),
            markers.get(BOTTOM_LEFT_ID).get(0)
        );
    }

    private void captureBottomMidpoints(Map<Integer, List<Point2D_F64>> markers) {
        bottomLeftTopOriginal = topMidpoint(markers.get(BOTTOM_LEFT_ID));
        bottomRightTopOriginal = topMidpoint(markers.get(BOTTOM_RIGHT_ID));
    }

    private Point2D_F64 topMidpoint(List<Point2D_F64> corners) {
        return new Point2D_F64((corners.get(0).x + corners.get(1).x) / 2.0, Math.min(corners.get(0).y, corners.get(1).y));
    }

    private Pair<Integer, Integer> computeOutputSize(List<Point2D_F64> srcPoints) {
        int width = (int) ((distance(srcPoints.get(0), srcPoints.get(1)) + distance(srcPoints.get(2), srcPoints.get(3))) / 2.0);
        int height = (int) ((distance(srcPoints.get(0), srcPoints.get(3)) + distance(srcPoints.get(1), srcPoints.get(2))) / 2.0);
        return new Pair<>(width, height);
    }

    private double distance(Point2D_F64 a, Point2D_F64 b) {
        return Math.hypot(a.x - b.x, a.y - b.y);
    }

    private Homography2D_F64 computeHomography(List<Point2D_F64> srcPoints, int width, int height) {
        List<Point2D_F64> dst = List.of(
            new Point2D_F64(0.0, 0.0),
            new Point2D_F64(width - 1.0, 0.0),
            new Point2D_F64(width - 1.0, height - 1.0),
            new Point2D_F64(0.0, height - 1.0)
        );
        List<AssociatedPair> pairs = new ArrayList<>();
        for (int i = 0; i < srcPoints.size(); i++) {
            pairs.add(new AssociatedPair(srcPoints.get(i), dst.get(i)));
        }
        DMatrixRMaj matrix = new DMatrixRMaj(3, 3);
        if (!new HomographyDirectLinearTransform(true).process(pairs, matrix)) {
            throw new RuntimeException("Failed to compute homography");
        }
        return UtilHomography_F64.convert(matrix, null);
    }

    private Planar<GrayF32> convertU8ToF32(Planar<GrayU8> src) {
        Planar<GrayF32> out = new Planar<>(GrayF32.class, src.width, src.height, 3);
        for (int band = 0; band < 3; band++) {
            GrayU8 sourceBand = src.getBand(band);
            GrayF32 destinationBand = out.getBand(band);
            for (int y = 0; y < src.height; y++) {
                for (int x = 0; x < src.width; x++) {
                    destinationBand.set(x, y, sourceBand.get(x, y));
                }
            }
        }
        return out;
    }

    private Planar<GrayF32> applyPerspective(Planar<GrayF32> srcF32, List<Point2D_F64> srcPoints, int width, int height) {
        RemovePerspectiveDistortion<Planar<GrayF32>> distortion = new RemovePerspectiveDistortion<>(width, height, ImageType.pl(3, GrayF32.class));
        if (!distortion.apply(srcF32, srcPoints.get(0), srcPoints.get(1), srcPoints.get(2), srcPoints.get(3))) {
            throw new RuntimeException("Failed to remove perspective");
        }
        return distortion.getOutput();
    }

    private Planar<GrayU8> convertF32ToU8(Planar<GrayF32> srcF32, int width, int height) {
        Planar<GrayU8> out = new Planar<>(GrayU8.class, width, height, 3);
        for (int band = 0; band < 3; band++) {
            GrayF32 sourceBand = srcF32.getBand(band);
            GrayU8 destinationBand = out.getBand(band);
            for (int y = 0; y < height; y++) {
                for (int x = 0; x < width; x++) {
                    int value = (int) sourceBand.get(x, y);
                    destinationBand.set(x, y, Math.max(0, Math.min(255, value)));
                }
            }
        }
        return out;
    }
}

File: test-utils/src/main/java/hu/kdea/szavazas/ballotprocessor/aruco/BoofCVArucoDetector.kt
package hu.kdea.szavazas.ballotprocessor

import boofcv.abst.fiducial.FiducialDetector
import boofcv.alg.distort.RemovePerspectiveDistortion
import boofcv.alg.geo.h.HomographyDirectLinearTransform
import boofcv.factory.fiducial.ConfigHammingMarker
import boofcv.factory.fiducial.FactoryFiducial
import boofcv.factory.fiducial.HammingDictionary
import boofcv.struct.geo.AssociatedPair
import boofcv.struct.image.GrayF32
import boofcv.struct.image.GrayU8
import boofcv.struct.image.ImageType
import boofcv.struct.image.Planar
import georegression.struct.homography.Homography2D_F64
import georegression.struct.homography.UtilHomography_F64
import georegression.struct.point.Point2D_F64
import georegression.struct.shapes.Polygon2D_F64
import georegression.transform.homography.HomographyPointOps_F64
import org.ejml.data.DMatrixRMaj
import kotlin.math.hypot
import kotlin.math.min

class BoofCVArucoDetector : IArucoDetector {

    private companion object {
        const val TOP_LEFT_ID = 37
        const val TOP_RIGHT_ID = 50
        const val BOTTOM_LEFT_ID = 44
        const val BOTTOM_RIGHT_ID = 219
        val REQUIRED_IDS = setOf(TOP_LEFT_ID, TOP_RIGHT_ID, BOTTOM_LEFT_ID, BOTTOM_RIGHT_ID)
    }

    private var bottomLeftTopOriginal: Point2D_F64? = null
    private var bottomRightTopOriginal: Point2D_F64? = null
    private var homography: Homography2D_F64? = null

    override fun findBallotCorners(gray: GrayU8): List<Point2D_F64>? {
        val markers = scanMarkers(gray) ?: return null
        captureBottomMidpoints(markers)
        Logger.d("ArUco", "Ballot corners detected successfully (IDs: ${markers.keys})")
        return extractOuterCorners(markers)
    }

    override fun warpBallot(src: Planar<GrayU8>, srcPoints: List<Point2D_F64>): Planar<GrayU8> {
        val (width, height) = computeOutputSize(srcPoints)
        homography = computeHomography(srcPoints, width, height)
        val srcF32 = convertU8ToF32(src)
        val outF32 = applyPerspective(srcF32, srcPoints, width, height)
        return convertF32ToU8(outF32, width, height)
    }

    override fun bottomMarkerTopInScaled(scaleFactor: Double): Double? {
        val h = homography ?: return null
        val bl = bottomLeftTopOriginal ?: return null
        val br = bottomRightTopOriginal ?: return null
        val ty1 = HomographyPointOps_F64.transform(h, bl, null).y
        val ty2 = HomographyPointOps_F64.transform(h, br, null).y
        return minOf(ty1, ty2) * scaleFactor
    }

    private fun scanMarkers(gray: GrayU8): Map<Int, List<Point2D_F64>>? {
        val detector = createDetector()
        detector.detect(gray)
        Logger.d("ArUco", "Total markers found: ${detector.totalFound()}")
        val markers = mutableMapOf<Int, List<Point2D_F64>>()
        for (i in 0 until detector.totalFound()) {
            val (id, corners) = readMarker(detector, i)
            logMarker(id, corners)
            if (id in REQUIRED_IDS && markers.containsKey(id)) {
                Logger.d("ArUco", "Duplicate required marker id $id – failing")
                return null
            }
            markers[id] = corners
        }
        if (!markers.keys.containsAll(REQUIRED_IDS)) {
            Logger.d("ArUco", "Missing required markers. Found IDs: ${markers.keys}")
            return null
        }
        return markers
    }

    private fun createDetector(): FiducialDetector<GrayU8> {
        val config = ConfigHammingMarker.loadDictionary(HammingDictionary.ARUCO_MIP_16h3)
        return FactoryFiducial.squareHamming(config, null, GrayU8::class.java)
    }

    private fun readMarker(
        detector: FiducialDetector<GrayU8>, i: Int
    ): Pair<Int, List<Point2D_F64>> {
        val id = detector.getId(i).toInt()
        val bounds = Polygon2D_F64()
        detector.getBounds(i, bounds)
        val corners = (0 until bounds.size()).map { bounds.get(it) }
        return id to corners
    }

    private fun logMarker(id: Int, corners: List<Point2D_F64>) {
        val cx = corners.map { it.x }.average()
        val cy = corners.map { it.y }.average()
        Logger.d("ArUco", "  id=$id  center=(${cx.toInt()},${cy.toInt()})")
    }

    private fun extractOuterCorners(markers: Map<Int, List<Point2D_F64>>) = listOf(
        markers[TOP_LEFT_ID]!![1],
        markers[TOP_RIGHT_ID]!![2],
        markers[BOTTOM_RIGHT_ID]!![3],
        markers[BOTTOM_LEFT_ID]!![0]
    )

    private fun captureBottomMidpoints(markers: Map<Int, List<Point2D_F64>>) {
        bottomLeftTopOriginal = topMidpoint(markers[BOTTOM_LEFT_ID]!!)
        bottomRightTopOriginal = topMidpoint(markers[BOTTOM_RIGHT_ID]!!)
    }

    private fun topMidpoint(corners: List<Point2D_F64>): Point2D_F64 =
        Point2D_F64((corners[0].x + corners[1].x) / 2.0, min(corners[0].y, corners[1].y))

    private fun computeOutputSize(srcPoints: List<Point2D_F64>): Pair<Int, Int> {
        fun d(a: Point2D_F64, b: Point2D_F64) = hypot(a.x - b.x, a.y - b.y)
        val width = ((d(srcPoints[0], srcPoints[1]) + d(srcPoints[2], srcPoints[3])) / 2).toInt()
        val height = ((d(srcPoints[0], srcPoints[3]) + d(srcPoints[1], srcPoints[2])) / 2).toInt()
        return width to height
    }

    private fun computeHomography(
        srcPoints: List<Point2D_F64>, width: Int, height: Int
    ): Homography2D_F64 {
        val dst = listOf(
            Point2D_F64(0.0, 0.0),
            Point2D_F64((width - 1).toDouble(), 0.0),
            Point2D_F64((width - 1).toDouble(), (height - 1).toDouble()),
            Point2D_F64(0.0, (height - 1).toDouble())
        )
        val pairs = srcPoints.zip(dst).map { (s, d) -> AssociatedPair(s, d) }
        val matrix = DMatrixRMaj(3, 3)
        if (!HomographyDirectLinearTransform(true).process(pairs, matrix))
            throw RuntimeException("Failed to compute homography")
        return UtilHomography_F64.convert(matrix, null)
    }

    private fun convertU8ToF32(src: Planar<GrayU8>): Planar<GrayF32> {
        val out = Planar(GrayF32::class.java, src.width, src.height, 3)
        for (band in 0 until 3) {
            val s = src.getBand(band); val d = out.getBand(band)
            for (y in 0 until src.height) for (x in 0 until src.width)
                d.set(x, y, s.get(x, y).toFloat())
        }
        return out
    }

    private fun applyPerspective(
        srcF32: Planar<GrayF32>, srcPoints: List<Point2D_F64>, width: Int, height: Int
    ): Planar<GrayF32> {
        val rp = RemovePerspectiveDistortion<Planar<GrayF32>>(
            width, height, ImageType.pl(3, GrayF32::class.java)
        )
        if (!rp.apply(srcF32, srcPoints[0], srcPoints[1], srcPoints[2], srcPoints[3]))
            throw RuntimeException("Failed to remove perspective")
        return rp.getOutput()
    }

    private fun convertF32ToU8(srcF32: Planar<GrayF32>, width: Int, height: Int): Planar<GrayU8> {
        val out = Planar(GrayU8::class.java, width, height, 3)
        for (band in 0 until 3) {
            val s = srcF32.getBand(band); val d = out.getBand(band)
            for (y in 0 until height) for (x in 0 until width)
                d.set(x, y, s.get(x, y).toInt().coerceIn(0, 255))
        }
        return out
    }
}